pandas/io/parsers.py

```diff
@@ -2346,9 +2346,17 @@ def __init__(self, f: Union[FilePathOrBuffer, List], **kwds):
             raise ValueError("Only length-1 decimal markers supported")
 
         if self.thousands is None:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.decimal}]+")
+            regex = fr"^\-?[0-9]*({self.decimal}[0-9]*)?([0-9](E|e)\-?[0-9]*)?$"
+            self.nonnum = re.compile(regex)
         else:
-            self.nonnum = re.compile(fr"[^-^0-9^{self.thousands}^{self.decimal}]+")
+            thousands = self.thousands
+            if "." == thousands:
+                thousands = fr"\{thousands}"
+            regex = (
+                fr"^\-?([0-9]+{thousands}|[0-9])*({self.decimal}[0-9]*)?"
```

Comment:
might be out of scope, but do we want to restrict to 3 digits between thousands characters (IIRC in some locales they separate on ten-thousands, not sure if we support that)

Reply:
That was a thing I was wondering too when opening this PR. Not quite sure how strict we would want to be here. We could maybe check for at least three digits between? The old version did not check anything related to this